### ISSUE TITLE
Connection failed to Kafka: p_kafka_check_outq_len() / p_kafka_close()

### ISSUE THREAD

**Issue by JonasGroeger** [pmacct/pmacct#34]
Hi!

I am running a self-compiled version of `nfactd` (1.6.0) on my server:

```
root@myserver:~ # nfacctd -V
NetFlow Accounting Daemon, nfacctd 1.6.0 (20160607-00)
 '--enable-kafka' '--enable-jansson' '--enable-ipv6'

For suggestions, critics, bugs, contact me: Paolo Lucente <paolo@pmacct.net>.
```

When starting it with a configuration like this:

```
nfacctd_ip: <ip>
nfacctd_port: <port>
interface: <interface>

!
! Configuration for Apache Kafka
!
plugins: kafka
aggregate: src_mac, dst_mac, vlan, cos, etype, src_host, dst_host, src_net, dst_net, src_mask, dst_mask, src_as, dst_as, src_port, dst_port, tos, proto, none, flows, tag, tag2, label, class, tcpflags, in_iface, out_iface, ext_comm, as_path, peer_src_ip, peer_dst_ip, peer_src_as, peer_dst_as, local_pref, med, src_ext_comm, src_as_path, src_local_pref, src_med, mpls_vpn_rd, mpls_label_top, mpls_label_bottom, mpls_stack_depth, sampling_rate, src_host_country, dst_host_country, nat_event, post_nat_src_host, post_nat_dst_host, post_nat_src_port, post_nat_dst_port, timestamp_start, timestamp_end, timestamp_arrival, export_proto_seqno, export_proto_version
kafka_topic: ipfix
kafka_broker_host: <server>
kafka_broker_port: <port>
kafka_refresh_time: 1
```
I get the following output:

```
root@myserver:~ # nfacctd -f /etc/pmacct/conf.conf 
INFO ( default/core ): Reading configuration file '/etc/pmacct/conf.conf'.
INFO ( default/core ): waiting for NetFlow data on <IP>:<port>
INFO ( default/kafka ): cache entries=16411 base cache memory=53434216 bytes
INFO ( default/kafka ): *** Purging cache - START (PID: 97597) ***
INFO ( default/kafka ): *** Purging cache - END (PID: 97597, QN: 0/0, ET: 0) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97600) ***
INFO ( default/kafka ): *** Purging cache - END (PID: 97600, QN: 0/0, ET: 0) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97603) ***
INFO ( default/kafka ): *** Purging cache - END (PID: 97603, QN: 0/0, ET: 0) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97606) ***
INFO ( default/kafka ): *** Purging cache - END (PID: 97606, QN: 0/0, ET: 0) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97609) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97614) ***
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_check_outq_len()
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_close()
INFO ( default/kafka ): *** Purging cache - END (PID: 97609, QN: 87/195, ET: 1) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97619) ***
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_check_outq_len()
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_close()
INFO ( default/kafka ): *** Purging cache - END (PID: 97614, QN: 52/120, ET: 1) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97624) ***
INFO ( default/kafka ): *** Purging cache - END (PID: 97619, QN: 106/106, ET: 2) ***
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_check_outq_len()
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_close()
INFO ( default/kafka ): *** Purging cache - END (PID: 97624, QN: 68/127, ET: 1) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97629) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97634) ***
^C
INFO ( default/kafka ): *** Purging cache - START (PID: 97634) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97629) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 97596) ***
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_check_outq_len()
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_close()
INFO ( default/kafka ): *** Purging cache - END (PID: 97596, QN: 78/130, ET: 1) ***
INFO ( default/kafka ): *** Purging cache - END (PID: 97634, QN: 259/259, ET: 2) ***
INFO ( default/core ): OK, Exiting ...
```

_Note: I also get those errors when setting `kafka_refresh_time` to something higher, like 30._

It seems like the plugin cannot connect to Kafka. However, the Kafka broker is up and running stable. Is this something I should worry about? How would I go about investigating this any further?

Apparently this comes from

* https://github.com/pmacct/pmacct/blob/master/src/kafka_common.c#L309
* https://github.com/pmacct/pmacct/blob/master/src/kafka_common.c#L314

When running `nfacctd` with `-d` in Debug mode, these errors do not show up :astonished:

**Comment by paololucente** [comment]
Hi Jonas, can you verify the same happens with code in master? Also, what happens if you set a higher kafka_refresh_time, say, to 60 secs? Are you in a controlled environment where you can expect what data is sent and what is effectively received by the consumer? In general, from the log above, it appears troubles start when multiple writers try to write to the broker in parallel.

**Comment by JonasGroeger** [comment]
> Hi Jonas, can you verify the same happens with code in master?

I tried the master branch and I'm running in the same errors.

> Also, what happens if you set a higher kafka_refresh_time, say, to 60 secs?

I tried the master branch with `kafka_refresh_time: 60` set. Same error.

> Are you in a controlled environment where you can expect what data is sent and what is effectively received by the consumer?

I'm in a controlled environment in the sense that I can control both producer and consumer, yes. But obviously the data I produce in `nfacctd` is in IPFIX format. I am receiving about 1000 messages / second.

> In general, from the log above, it appears troubles start when multiple writers try to write to the broker in parallel.

Can I limit this to 1 producer somehow? Since librdkafka should handle my workload easily.

**Comment by JonasGroeger** [comment]
I tried setting `kafka_max_writers: 1`. No messages are lost.
The default for `kafka_max_writers` is `10`. Even with this setting (and the errors) no messages are lost.

(So pmacct / kafka plugin CAN indeed correctly connect to the kafka service and the *could not connect* issue is something else?)

**Comment by paololucente** [comment]
Hi Jonas, you anticipated me on setting that kafka_max_writers to 1, perfect. Thanks also for all the troubleshooting: it all points to my initial suspicion that the error messages are false positives and are spit when multiple writers are writing. Let me reproduce in lab and fix the issue.

**Comment by paololucente** [comment]
Hi Jonas, i did try to reproduce this in lab - unsuccessfully. I have to admit i had to artificially re-create my understanding of your scenario. Any chance i can access your setup so to see what is going on? If yes, please get in touch privately by email. If not, i'm afraid i'm unable to help you: i will keep the issue open bug have to mark it as "wontfix".

**Comment by JonasGroeger** [comment]
I'll come back to you on tuesday.

**Comment by paololucente** [comment]
Hi Jonas, was wondering whether you had any news on this? If not, i'd either close (we can re-open) or mark as wontfix until we can progress on it.

**Comment by viktorchukhantsev** [comment]
Such problems appear when kafka configured wrong. Like you have zookeeper and several kafka instances with incorrect ip addresses in configs.

**Comment by JonasGroeger** [comment]
> Any chance i can access your setup so to see what is going on?

Since I'm in a corporate environment, I'm sorry but I don't think this is possible.

@viktorchukhantsev Which configuration are you talking about? In `pmacct` I do not have to set the Zookeeper hostnames, only the Kafka hostnames, like above:

```
nfacctd_ip: 1.2.3.4
nfacctd_port: 2100
interface: eth0

!
! Configuration for Apache Kafka
!
plugins: kafka
aggregate: src_mac, dst_mac, …
kafka_topic: ipfix
kafka_broker_host: server01.company.com:9092,server02.company.com:9092,server03.company.com:9092
kafka_refresh_time: 10
```

**Comment by JonasGroeger** [comment]
I'm still getting these errors but since I'm not losing data, its fine.

**Comment by viktorchukhantsev** [comment]
@JonasGroeger I mean this problem is related to kafka more than to pmacct.

**Comment by cpmarvin** [comment]
hi.
running latest git code i'm still getting these errors.

ERROR ( ukb/kafka ): Connection failed to Kafka: p_kafka_check_outq_len()
Aug 05 17:10:03 ERROR ( ukb/kafka ): Connection failed to Kafka: p_kafka_close()

- librdkafka 0.8.3

**Comment by paololucente** [comment]
@cpmarvin : can you please describe the issue better? Are these messages resulting in loss of data? Is this happening one-off only, ie. when the topic is not created yet?

**Comment by cpmarvin** [comment]
hi Paolo,

Happens everytime after initial startup, and results in lost data.  The topic is created before startup.

pluging conf: 

kafka_topic[ukb]: acct
kafka_refresh_time[ukb]: 300
kafka_history[ukb]: 5m
kafka_history_roundoff[ukb]: m
kafka_max_writers[ukb]: 1

Regards,
Catalin

**Comment by cpmarvin** [comment]
hi Paolo,

Happens everytime after initial startup, and results in lost data. The
topic is created before startup.

pluging conf:

kafka_topic[ukb]: acct
kafka_refresh_time[ukb]: 300
kafka_history[ukb]: 5m
kafka_history_roundoff[ukb]: m
kafka_max_writers[ukb]: 1

Let me know if you need more info.


Thx.

**Comment by paololucente** [comment]
Hi Catalin,

Thanks for the extra insight. I tried with your exact same config without success (ie. all works well). My test environment is:

* librdkafka 0.9.1
* Kafka 0.8.2.2
* Scala 2.9.2

Any chance you can refresh your librdkafka and do another check?

**Comment by cpmarvin** [comment]
Hi Paolo,

with kafka 0.8.2.2 (kafka_2.9.2-0.8.2.2-scaladoc.jar) and librdkafka 0.9.X

Aug 08 00:05:04 ERROR ( ukb/kafka ): Kafka message delivery failed: Local:
Bad message format
Aug 08 00:05:04 ERROR ( ukb/kafka ): Kafka message delivery failed: Local:
Bad message format
Aug 08 00:05:04 ERROR ( ukb/kafka ): Kafka message delivery failed: Local:
Bad message format
Aug 08 00:05:05 INFO ( ukb/kafka ): *** Purging cache - END (PID: 25488,
QN: 124/124, ET: 0) ***

The topic gets updated with data , but i need to double check that are
correct. From the looks of it there are couple of missing flows but i need
more time to double check.

Regards,
Catalin

**Comment by cpmarvin** [comment]
Hi Paolo,

Aug 08 01:45:02 DEBUG ( ukb/kafka ): Kafka message delivery successful

using librdkafka-0.9.0.

Hope this helps.

Catalin

**Comment by JonasGroeger** [comment]
It seems we need to set `api.version.request=true` when compiling?

**Comment by paololucente** [comment]
Hi @JonasGroeger , @cpmarvin 

I've just tried to refresh my environment to see if i could reproduce the issue (or, like @JonasGroeger was suggesting, some extra parameter is to be set up). All seems to work well for me both in the case of writing to an existing topic or to a new topic, without any Kafka tweaking (like: download, install and run according to their quickstart document), also with this combination:

* pmacct >= 20160729-00
* librdkafka 0.9.1
* Kafka 0.10.0
* Scala 2.11.8

These are latest stable of the three, any chance you guys can check this combination and see if it works for you too? I'd like to verify whether it's just an issue with compatibility of the components (which would trigger just an update of the documentation at my side) or there is something else going on.

Should things not work, it would be much appreciated if you can share your Kafka configs/tweaks, if any (as i already have both your pmacct configs; they look pretty basic and i've already tested both of them working).

Paolo

**Comment by juliankeppel** [comment]
I get the same error with the following set up:

pmacct 1.6.0
librdkafka 0.8.6-1.1 (from latest Ubuntu Package)
Kafka 0.10.0.1
Scala 2.11

I run in a Kubernetes/Google Container Engine setup. The pmacct forwards all packets to the Kafka broker (I checked this with tcpdump). And if I configure a topic which isn`t there already, Kafka adds the topic (which I can check in the log folders).

But the events don't get into the topic log and are not receivable with the console-consumer. In the logs I found the same error messages as JonasGroeger: 

ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_check_outq_len()
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_close()
INFO ( default/kafka ): *** Purging cache - END (PID: 1141, QN: 0/2016, ET: 1) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 1145) ***
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_check_outq_len()
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_close()
INFO ( default/kafka ): *** Purging cache - END (PID: 1145, QN: 0/2461, ET: 1) ***
INFO ( default/kafka ): *** Purging cache - START (PID: 1149) ***
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_check_outq_len()
ERROR ( default/kafka ): Connection failed to Kafka: p_kafka_close()
INFO ( default/kafka ): *** Purging cache - END (PID: 1149, QN: 0/2606, ET: 1) ***
...
...

I also tweaked the refresh time up, without any success.

**Comment by paololucente** [comment]
Hi Julian,

From the QUICKSTART doc, two combinations have been tested working:

* librdkafka 0.9.1, Kafka 0.8.2.2 and Scala 2.9.2
* librdkafka 0.9.1, Kafka 0.10.0 and Scala 2.11.8

Can you please refresh & self-compile your librdkafka? If yes, please keep me/us posted what is the outcome of that. Should that not work, any chance you can provide remote access to your system for some debugging/troubleshooting?

Thanks,
Paolo

**Comment by juliankeppel** [comment]
Hi Paolo,

I now tried to self-compile librdkafka from master branch. The same error appears and no events are written to the kafka log.

I'm sorry but I can't give you access to the environment because it't a Google Cloud Project from my company. 

It seems like it's the same problem as Jonas has had... the connection itself seems to work properly.

**Comment by paololucente** [comment]
Hi Julian,

Sure, i understand. Unfortunately not having a way to reproduce it, i can't do much about this issue. If you by any chance manage to drill it down to anything (ie. amount of entries to dump, OS, version of some of the software, etc.) that can bring us closer to solution, please let me know.

**Comment by juliankeppel** [comment]
- The complete infrastructure runs on Google Container Engine with Kubernetes 1.3.7
- pmacct 1.6.0 runs on an Ubuntu 16.04 Docker Container
- I compiled librdkafka from master branch (I also tried 0.9.1-1 and 0.8.6-1.1)
- Kafka version 0.10.0.1 with Scala 2.11.8 (I also tried kafka 0.10.0 and 0.8.2.2)

I tried different settings of  kafka_max_writers and kafka_refresh_time without any success. I also tried to connect with IP and hostname, both with the same result. As I mentioned before, the interesting thing is that a new topic is auto created by kafka when I configure the topic to be a new one in pmacct.

@JonasGroeger: Do you have a solution for you problem now? What about api.version.request=true, as you mentioned earlier? Did that help? If yes, how can I configure librdkafka? I haven't found any documentation abount how to configure at https://github.com/edenhill/librdkafka yet.

**Comment by juliankeppel** [comment]
I solved the problem. It was a problem with the kafka configuration, so it wasn't any problem with pmacct. 

In default setting, kafka only listens to local connections. I changed the corresponding settings with sed in my Dockerfile. But this seemed not to work properly. Now I provide a prepared configuration file for kafka and use this in my Dockerfile. The only confusing thing is that the initial topic creation worked though...

Thanks for you help.

**Comment by paololucente** [comment]
Hi Julian,

This is fantastic. Since it appears the problem happened to multiple people, ie. only in this thread three, do you think you can transform the above statement in a FAQ or piece of doc to embed in the QUICKSTART doc? If not, i may do that no problem but can you be slightly more specific on the description of what did work, what did not and what was confusing? Thanks in advance for any effort you can spare on this.

Cheers,
Paolo

**Comment by juliankeppel** [comment]
Yes I could do this. 

But I think the problem was more a topic of **Dockerized Kafka** and in the end it really hasn't had something to do with pmacct or librdkafka. One can find how to configure a Kafka broker in a cloud environment there: http://kafka.apache.org/documentation.html#brokerconfigs

The only problem was to set the necessary configuration parameters in the Docker container. And I think this is really a Docker-specific topic. What do you think?

**Comment by paololucente** [comment]
Hi Julian,

Thanks for the link and the extra information. I guess i will then just add a note in the QUICKSTART doc and link this GitHub issue for anybody stumbling upon the same issue.

Thanks,
Paolo

**Comment by kenzodeluxe** [comment]
For those stumbling upon 
> Connection failed to Kafka: p_kafka_check_outq_len()

and don't appear to have obvious Kafka Docker/configuration issues, there's at least one more way of running into this when using pmacct in a released version <= 1.7.1:
When producing the message hasn't led to a decrease in queue length, the released code will retry once and then fail (first change in commit bef67c56a6794222992e48f62cbc95d34421e74a, second in 2b13a83fd875858d44581440ace3e696d630c182). Depending on latency from pmacct to Kafka, it may take a little longer to actually get the message produced; after the change, PM_KAFKA_OUTQ_LEN_RETRIES takes care of that. So if you've run into the same issue, try master first to see if it fixed it for you.
Something like this was supposed to be my first PR for pmacct, but I was too late! ;)
